File: backend/app/api/problems.py
from fastapi import APIRouter, HTTPException, status
from typing import List, Optional
from datetime import datetime
from bson import ObjectId
from ..models.problem import Problem, ProblemCreate, ProblemUpdate
from ..core import database

router = APIRouter(prefix="/problems", tags=["problems"])

def problem_helper(problem) -> dict:
    """Convert MongoDB document to dictionary"""
    def convert_objectid(obj):
        """Recursively convert ObjectId to string in nested structures"""
        if isinstance(obj, ObjectId):
            return str(obj)
        elif isinstance(obj, dict):
            return {k: convert_objectid(v) for k, v in obj.items()}
        elif isinstance(obj, list):
            return [convert_objectid(item) for item in obj]
        else:
            return obj
    
    return {
        "id": str(problem["_id"]),
        "title": problem.get("title", ""),
        "difficulty": problem.get("difficulty", "Easy"),
        "tags": convert_objectid(problem.get("tags", [])),
        "companies": convert_objectid(problem.get("companies", [])),
        "description": problem.get("description", ""),
        "examples": convert_objectid(problem.get("examples", [])),
        "constraints": convert_objectid(problem.get("constraints", [])),
        "testCases": convert_objectid(problem.get("testCases", [])),
        "codeTemplates": convert_objectid(problem.get("codeTemplates", {})),
        "functionSignatures": convert_objectid(problem.get("functionSignatures")),
        "stats": convert_objectid(problem.get("stats", {})),
        "published": problem.get("published", False),
        "createdAt": problem.get("createdAt"),
        "updatedAt": problem.get("updatedAt")
    }

@router.get("/")
async def get_all_problems(published_only: Optional[bool] = None):
    """Get all problems (admin) or only published problems (candidates)"""
    try:
        from ..core.database import get_mongo_db
        
        mongodb = get_mongo_db()
        
        if mongodb is None:
            raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                detail="Database not available"
            )
        
        problems_coll = mongodb["problems"]
        
        query = {}
        if published_only:
            query["published"] = True
        
        problems = []
        async for problem in problems_coll.find(query):
            problems.append(problem_helper(problem))
        
        return problems
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error fetching problems: {str(e)}"
        )

@router.get("/{problem_id}")
async def get_problem(problem_id: str):
    """Get a single problem by ID"""
    try:
        from ..core.database import get_mongo_db
        
        mongodb = get_mongo_db()
        
        if not ObjectId.is_valid(problem_id):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid problem ID"
            )
        
        problems_coll = mongodb["problems"]
        problem = await problems_coll.find_one({"_id": ObjectId(problem_id)})
        if not problem:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Problem not found"
            )
        
        return problem_helper(problem)
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error fetching problem: {str(e)}"
        )

@router.post("/", status_code=status.HTTP_201_CREATED)
async def create_problem(problem: ProblemCreate):
    """Create a new problem (admin only)"""
    try:
        from ..core.database import get_mongo_db
        
        mongodb = get_mongo_db()
        problems_coll = mongodb["problems"]
        problem_dict = problem.dict()
        problem_dict["createdAt"] = datetime.utcnow()
        problem_dict["updatedAt"] = datetime.utcnow()
        
        result = await problems_coll.insert_one(problem_dict)
        created_problem = await problems_coll.find_one({"_id": result.inserted_id})
        
        return problem_helper(created_problem)
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error creating problem: {str(e)}"
        )

@router.put("/{problem_id}")
async def update_problem(problem_id: str, problem_update: ProblemUpdate):
    """Update a problem (admin only)"""
    try:
        from ..core.database import get_mongo_db
        
        mongodb = get_mongo_db()
        
        if not ObjectId.is_valid(problem_id):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid problem ID"
            )
        
        # Get existing problem
        existing_problem = await mongodb["problems"].find_one({"_id": ObjectId(problem_id)})
        if not existing_problem:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Problem not found"
            )
        
        # Update only provided fields
        update_dict = problem_update.dict(exclude_unset=True)
        update_dict["updatedAt"] = datetime.utcnow()
        
        await mongodb["problems"].update_one(
            {"_id": ObjectId(problem_id)},
            {"$set": update_dict}
        )
        
        updated_problem = await mongodb["problems"].find_one({"_id": ObjectId(problem_id)})
        return problem_helper(updated_problem)
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error updating problem: {str(e)}"
        )

@router.patch("/{problem_id}/publish")
async def toggle_publish(problem_id: str):
    """Toggle publish status of a problem"""
    try:
        from ..core.database import get_mongo_db
        
        mongodb = get_mongo_db()
        
        if not ObjectId.is_valid(problem_id):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid problem ID"
            )
        
        problem = await mongodb["problems"].find_one({"_id": ObjectId(problem_id)})
        if not problem:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Problem not found"
            )
        
        new_status = not problem.get("published", False)
        
        await mongodb["problems"].update_one(
            {"_id": ObjectId(problem_id)},
            {"$set": {"published": new_status, "updatedAt": datetime.utcnow()}}
        )
        
        updated_problem = await mongodb["problems"].find_one({"_id": ObjectId(problem_id)})
        return problem_helper(updated_problem)
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error toggling publish status: {str(e)}"
        )

@router.delete("/{problem_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_problem(problem_id: str):
    """Delete a problem (admin only)"""
    try:
        from ..core.database import get_mongo_db
        
        mongodb = get_mongo_db()
        
        if not ObjectId.is_valid(problem_id):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid problem ID"
            )
        
        result = await mongodb["problems"].delete_one({"_id": ObjectId(problem_id)})
        if result.deleted_count == 0:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Problem not found"
            )
        
        return None
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error deleting problem: {str(e)}"
        )

File: server/routes/problems.js
import express from 'express';
import Problem from '../models/Problem.js';

const router = express.Router();

// Get all problems
router.get('/', async (req, res) => {
  try {
    const { published_only } = req.query;
    const query = published_only ? { published: true } : {};
    const problems = await Problem.find(query).sort({ serialNumber: 1 });
    // Map _id to id for frontend compatibility
    const mappedProblems = problems.map(p => ({
      ...p.toObject(),
      id: p._id.toString()
    }));
    res.json(mappedProblems);
  } catch (error) {
    console.error('Error fetching problems:', error);
    res.status(500).json({ success: false, message: 'Error fetching problems' });
  }
});

// Get single problem by ID
router.get('/:id', async (req, res) => {
  try {
    const problem = await Problem.findById(req.params.id);
    if (!problem) {
      return res.status(404).json({ success: false, message: 'Problem not found' });
    }
    // Map _id to id for frontend compatibility
    const mappedProblem = {
      ...problem.toObject(),
      id: problem._id.toString()
    };
    res.json(mappedProblem);
  } catch (error) {
    console.error('Error fetching problem:', error);
    res.status(500).json({ success: false, message: 'Error fetching problem' });
  }
});

// Create new problem
router.post('/', async (req, res) => {
  try {
    // Get the highest serial number and increment by 1
    const lastProblem = await Problem.findOne().sort({ serialNumber: -1 });
    const nextSerialNumber = lastProblem ? lastProblem.serialNumber + 1 : 1;

    const problem = new Problem({
      ...req.body,
      serialNumber: nextSerialNumber
    });
    await problem.save();
    // Map _id to id for frontend compatibility
    const mappedProblem = {
      ...problem.toObject(),
      id: problem._id.toString()
    };
    res.json(mappedProblem);
  } catch (error) {
    console.error('Error creating problem:', error);
    if (error.name === 'ValidationError') {
      return res.status(400).json({
        success: false,
        message: 'Validation failed',
        errors: Object.keys(error.errors).map(field => ({
          field,
          message: error.errors[field].message
        }))
      });
    }
    res.status(500).json({ success: false, message: 'Error creating problem' });
  }
});

// Update problem
router.put('/:id', async (req, res) => {
  try {
    const problem = await Problem.findByIdAndUpdate(
      req.params.id,
      { $set: req.body },
      { new: true, runValidators: true }
    );
    if (!problem) {
      return res.status(404).json({ success: false, message: 'Problem not found' });
    }
    // Map _id to id for frontend compatibility
    const mappedProblem = {
      ...problem.toObject(),
      id: problem._id.toString()
    };
    res.json(mappedProblem);
  } catch (error) {
    console.error('Error updating problem:', error);
    res.status(500).json({ success: false, message: 'Error updating problem' });
  }
});

// Toggle publish status
router.patch('/:id/publish', async (req, res) => {
  try {
    const problem = await Problem.findById(req.params.id);
    if (!problem) {
      return res.status(404).json({ success: false, message: 'Problem not found' });
    }
    problem.published = !problem.published;
    await problem.save();
    // Map _id to id for frontend compatibility
    const mappedProblem = {
      ...problem.toObject(),
      id: problem._id.toString()
    };
    res.json(mappedProblem);
  } catch (error) {
    console.error('Error toggling publish status:', error);
    res.status(500).json({ success: false, message: 'Error toggling publish status' });
  }
});

// Delete problem
router.delete('/:id', async (req, res) => {
  try {
    const problem = await Problem.findByIdAndDelete(req.params.id);
    if (!problem) {
      return res.status(404).json({ success: false, message: 'Problem not found' });
    }
    res.status(204).send();
  } catch (error) {
    console.error('Error deleting problem:', error);
    res.status(500).json({ success: false, message: 'Error deleting problem' });
  }
});

export default router;
